grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,256 @@
+                throw new IllegalStateException("Cannot be added to more than one server");
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+        server.addListener(new ServerListenerAdapter() {
+            @Override
+            public void serverStarted(Server server) {
+                serverHealth.setHealthy(true);
+            }
+
+            @Override
+            public void serverStopping(Server server) {
+                serverHealth.setHealthy(false);
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(@Nullable String serverName) {
+        if (!isHealthy()) {
+            return ServingStatus.NOT_SERVING;
+        }
+        if (Strings.isNullOrEmpty(serverName)) {
+            if (grpcServiceHealthCheckers.isEmpty()) {
+                return ServingStatus.SERVING;
+            }
+            if (grpcServiceHealthCheckers.values().stream().allMatch(HealthChecker::isHealthy)) {
+                return ServingStatus.SERVING;
+            } else {
+                return ServingStatus.NOT_SERVING;
+            }
+        } else {
+            if (grpcServiceHealthCheckers.containsKey(serverName)) {
+                if (grpcServiceHealthCheckers.get(serverName).isHealthy()) {
+                    return ServingStatus.SERVING;
+                } else {
+                    return ServingStatus.NOT_SERVING;
+                }
+            } else {
+                return ServingStatus.SERVICE_UNKNOWN;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    void changeServerStatus(boolean isHealthy) {
+        serverHealth.setHealthy(isHealthy);
+    }
+
+    private void addServerHealthUpdateListener(ImmutableList<HealthCheckUpdateListener> updateListeners) {
+        serverHealth.addListener(healthChecker -> {
+            updateListeners.forEach(updateListener -> {
+                try {
+                    updateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private void setInternalHealthUpdateListener(Collection<ListenableHealthChecker> listenableHealthCheckers) {
+        listenableHealthCheckers.forEach(lhc -> {
+            lhc.addListener(healthChecker -> {
+                try {
+                    provideInternalHealthUpdateListener().healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private HealthCheckUpdateListener provideInternalHealthUpdateListener() {
+        return isHealthy -> {
+            for (Entry<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>> entry
```

Comment:
If multiple listeners are updated simultaneously, duplicate iterator seems inefficient. 
How about adding a flag for indicating the updates?
```java
if (updating) {
    return;
}
updating = true;
try {

} finally {
    updating = false
}
```